### PR TITLE
Fleshed out CircuitInternal

### PR DESCRIPTION
**Pull request by KevinMackenzie**
- preliminary `applyOps` impl
- more front-end functions
- more flexible transactions
- `CircuitLog` proto-OT logic
- Low-level event propagation
- expanded "Info" types.

Please ignore the egregiously inconsistent Error strings for now.

### CODE REVIEW

File: src/app/core/internal/impl/CircuitInternal.ts [hunk 1]
```diff
@@ -1,187 +1,392 @@
...
+            // Sanity check: Clock should be kept updated by the event handler.
+            if (this.clock !== this.log.clock())
+                throw new Error("Unexpected clock difference.  Maybe a missed event?");
+            this.log.propose(txOps);
+        });
     }
-    public redo(): boolean {
-        return this.undoRedoHelper(this.history.redo());
+    public cancelTransaction(): void {
+        this.endTransactionHelper(() => {});
```
Comment (KevinMackenzie) [inline src/app/core/internal/impl/CircuitInternal.ts:247]: Whoops.  This needs to revert the `transactionOps`

File: src/app/core/internal/impl/CircuitInternal.ts [hunk 2]
```diff
@@ -1,187 +1,392 @@
...
     // Convenience functions around CircuitOps
     //
 
-    public placeComponent(kind: string, props: Schema.Component["props"]): void {
-        this.applyOp({
+    protected addTransactionOp(op: CircuitOp): void {
+        this.assertTransactionState(true);
+        this.transactionOps.push(op);
+        // read-your-writes
+        this.applyOp(op);
```
Comment (KevinMackenzie) [inline src/app/core/internal/impl/CircuitInternal.ts:264]: Cancel the transaction if `applyOp` throws an exception here

**Review by LeonMontealegre — CHANGES_REQUESTED** [review]

File: src/app/core/internal/impl/CircuitOps.ts [hunk 3]
```diff
@@ -1,18 +1,11 @@
...
-    ops: CircuitOp[];
-}
 
 export interface PlaceComponentOp {
     kind: "PlaceComponentOp";
     inverted: boolean;
     c: Schema.Component;
-    ports: Schema.Port[];
-    wires: Schema.Wire[];
+    // Components are born without ports
```
[the remark this answers is not in the document]
Comment (KevinMackenzie) [reply src/app/core/internal/impl/CircuitOps.ts:8]: It helps reduce the complexity of the individual ops and doesn't introduce any problematic intermediary states.  I'll add a comment to that effect.

File: src/app/core/internal/impl/CircuitInternal.ts [hunk 4]
```diff
@@ -1,187 +1,392 @@
...
+        this.addTransactionOp({
+            kind:     "ConnectWireOp",
+            inverted: true,
+            w:        obj, // No copy needed
+        })
     }
 
-    public deleteObject(id: GUID): void {
-        throw new Error("Unimplemented");
+    public setPropForDyn(id: GUID, key: string, newVal?: Schema.Prop): void {
```
[the remark this answers is not in the document]
Comment (KevinMackenzie) [reply src/app/core/internal/impl/CircuitInternal.ts:326]: Short for "Dynamic".  Will make name more clear.

File: src/app/core/public/api/impl/Component.ts [hunk 5]
```diff
@@ -36,14 +35,14 @@ export class ComponentImpl extends BaseObjectImpl implements Component {
...
+        this.circuit.setPropForDyn(this.id, "angle", val);
     }
     public get angle(): number {
         return (this.getObj().props.angle ?? 0);
     }
 
     public get ports(): Port[] {
-        return this.circuit.getPortsForComponent(this.getObj())
-            .map((p) => new PortImpl(this.circuit, p.id));
+        return [...this.circuit.getPortsForComponent(this.id)]
```
[the remark this answers is not in the document]
Comment (KevinMackenzie) [reply src/app/core/public/api/impl/Component.ts:45]: `Set` doesn't have a `.map`.

**Review by LeonMontealegre — APPROVED** [review]

**Pull request merged**